In [1]:
import pandas as pd 
import numpy as np 
import os 
import matplotlib.pyplot as plt
import seaborn as sns 
import datetime as dict

In [5]:
cwd = os.getcwd()
data = pd.read_excel(cwd+"/KPMG_VI_New_raw_data_update_final.xlsx",'Transactions',header = 1)

In [7]:
data.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [8]:
data.shape

(20000, 13)

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [11]:
data.duplicated().value_counts()

False    20000
Name: count, dtype: int64

In [12]:
data.transaction_date.dt.year.value_counts()

transaction_date
2017    20000
Name: count, dtype: int64

In [14]:
data.order_status.value_counts()

order_status
Approved     19821
Cancelled      179
Name: count, dtype: int64

Missing Values

In [15]:
data.online_order.value_counts()

online_order
1.0    9829
0.0    9811
Name: count, dtype: int64

In [16]:
data[(data.online_order != 1.0) &(data.online_order !=0.0 )]

transaction_id  product_id  customer_id transaction_date  online_order  \
97                 98          49          333       2017-06-23           NaN   
166               167          90         3177       2017-04-26           NaN   
169               170           6          404       2017-10-16           NaN   
250               251          63         1967       2017-04-11           NaN   
300               301          78         2530       2017-03-24           NaN   
...               ...         ...          ...              ...           ...   
19514           19515          51          690       2017-01-22           NaN   
19573           19574          18         1735       2017-01-15           NaN   
19580           19581          49         1933       2017-10-12           NaN   
19635           19636          98         1389       2017-07-26           NaN   
19843           19844           3            8       2017-01-28           NaN   

      order_status           brand product_line product_class product_size  \
97        Approved   Trek Bicycles         Road        medium       medium   
166       Approved  Norco Bicycles     Standard           low       medium   
169       Approved      OHM Cycles     Standard          high       medium   
250       Approved           Solex     Standard        medium       medium   
300       Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19514     Approved      OHM Cycles     Standard          high       medium   
19573     Approved           Solex     Standard        medium       medium   
19580     Approved   Trek Bicycles         Road        medium       medium   
19635     Approved   Trek Bicycles     Standard          high       medium   
19843     Approved   Trek Bicycles     Standard        medium        large   

       list_price  standard_cost  product_first_sold_date  
97         533.51         400.13                  37823.0  
166        363.01         290.41                  38482.0  
169        227.88         136.73                  37838.0  
250       1483.20          99.59                  42145.0  
300       1765.30         709.48                  35455.0  
...           ...            ...                      ...  
19514     2005.66        1203.40                  37823.0  
19573      575.27         431.45                  41345.0  
19580      533.51         400.13                  41064.0  
19635      358.39         215.03                  38002.0  
19843     2091.47         388.92                  37823.0  

[360 rows x 13 columns]

Online order columns has 360 missing values while filled entries are almost equally distributed. These missing values could also mean other modes of ordering or might even be the exchange orders. Let us just provide another floating value of 2.0 to the missing values indicating other modes of ordering. 

In [17]:
data.online_order = data.online_order.fillna(2.0)

In [18]:
data.online_order.value_counts()

online_order
1.0    9829
0.0    9811
2.0     360
Name: count, dtype: int64

### Missing values in Brand, product line, product class, product size, standard cost and product first sold date

All these columns have same amount of missing values as we saw from info data. Now we will see if these are missing for the same entries (rows) across these columns.

In [19]:
data[data.brand.isna()]

transaction_id  product_id  customer_id transaction_date  online_order  \
136               137           0          431       2017-09-23           0.0   
159               160           0         3300       2017-08-27           0.0   
366               367           0         1614       2017-03-10           0.0   
406               407           0         2559       2017-06-14           1.0   
676               677           0         2609       2017-07-02           0.0   
...               ...         ...          ...              ...           ...   
19340           19341           0          443       2017-12-26           1.0   
19383           19384           0         2407       2017-06-11           0.0   
19793           19794           0         2860       2017-01-13           0.0   
19859           19860           0         2468       2017-06-24           1.0   
19871           19872           0           61       2017-03-17           1.0   

      order_status brand product_line product_class product_size  list_price  \
136       Approved   NaN          NaN           NaN          NaN     1942.61   
159       Approved   NaN          NaN           NaN          NaN     1656.86   
366       Approved   NaN          NaN           NaN          NaN      850.89   
406       Approved   NaN          NaN           NaN          NaN      710.59   
676       Approved   NaN          NaN           NaN          NaN     1972.01   
...            ...   ...          ...           ...          ...         ...   
19340     Approved   NaN          NaN           NaN          NaN      744.54   
19383     Approved   NaN          NaN           NaN          NaN     1098.18   
19793     Approved   NaN          NaN           NaN          NaN      868.56   
19859     Approved   NaN          NaN           NaN          NaN     1497.43   
19871     Approved   NaN          NaN           NaN          NaN      867.92   

       standard_cost  product_first_sold_date  
136              NaN                      NaN  
159              NaN                      NaN  
366              NaN                      NaN  
406              NaN                      NaN  
676              NaN                      NaN  
...              ...                      ...  
19340            NaN                      NaN  
19383            NaN                      NaN  
19793            NaN                      NaN  
19859            NaN                      NaN  
19871            NaN                      NaN  

[197 rows x 13 columns]

In [20]:
data.product_line.value_counts()

product_line
Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: count, dtype: int64

In [21]:
data.product_class.value_counts()

product_class
medium    13826
high       3013
low        2964
Name: count, dtype: int64

In [22]:
data.product_size.value_counts()

product_size
medium    12990
large      3976
small      2837
Name: count, dtype: int64

In [23]:
data.standard_cost = data.standard_cost.fillna(data.standard_cost.mean())


We can see that there are 197 entries that is missing data about above mentioned columns. If this data is below 5 % of all data we can delete these rows from the table. However, we are only working to create visualization for now and not with predicting, so we will keep these null values. 